.github/workflows/build.yml

```diff
@@ -16,9 +16,9 @@ on:
 jobs:
   build:
     env:
-      TEST_TIMEOUT_MINUTES: 30
+      TEST_TIMEOUT_MINUTES: 40
       FSAC_TEST_DEFAULT_TIMEOUT : 120000 #ms, individual test timeouts
-    timeout-minutes: 30 # we have a locking issue, so cap the runs at ~20m to account for varying build times, etc
+    timeout-minutes: 40 # we have a locking issue, so cap the runs at ~20m to account for varying build times, etc
```

Comment:
it concerns me a tiny bit that we keep climbing on this :(

Reply:
I basically doubled the test suite with `NamedText`/`RoslynSourceText`. Might be worth switching to `RoslynSourceText` by default then eventually removing `NamedText`

Reply:
ok, so this isn't because individual tests are hanging, it's because the suite overall is going long? that's more understandable. I'd agree with both of the things you mention.

I'm guessing there's some Expecto cancellation weirdness that's preventing us from having a truly per-test timeout?

Reply:
> I'm guessing there's some Expecto cancellation weirdness that's preventing us from having a truly per-test timeout?

I think this does work (I see some tests fail at 2 mins) but we were hitting the overall timeout we set on the CI. I know I had to make a lot of tests sequential to keep the test suite passing consistently and that's a big part of it being slow too. Probably need to revisit the setup and see why things tend to not be thread safe. Might simply be we're not making an LSP server per test or something else. 

I do vaguely remember the "wait for parse/check notifications" is how a lot of tests know how to continue after requesting a typecheck which can fail and not provide the correct feedback or gets lost in the logging.

Reply:
> I'm guessing there's some Expecto cancellation weirdness that's preventing us from having a truly per-test timeout?

Yes, individual timeouts do work. It's "just" the overall timeout we run into.


> I basically doubled the test suite with NamedText/RoslynSourceText. Might be worth switching to RoslynSourceText by default then eventually removing NamedText

I think that's reasonable. They should act the same, so just using one should be enough (and then maybe some tests to ensure they really act the same). Currently we're practically doing all tests twice -- despite not actually testing the `XXXText`, but the server functionality.   
Locally the first thing I do before running tests is always uncomment `NamedText` -- which basically cuts the test time in halve.  
Other "multipliers" are already uncommented ([[1]](https://github.com/fsharp/FsAutoComplete/blob/f6d5ddcad9e82eb3cb52e01cd3c86d40a205dabe/test/FsAutoComplete.Tests.Lsp/Program.fs#L34-L38) [[2]](https://github.com/fsharp/FsAutoComplete/blob/f6d5ddcad9e82eb3cb52e01cd3c86d40a205dabe/test/FsAutoComplete.Tests.Lsp/Program.fs#L53-L57)).

------------------

> Probably need to revisit the setup and see why things tend to not be thread safe. 

I **think** it's just the initialization and destruction of LSP server that must be sequential. The tests themselves should be able to run in parallel -- but are currently in [sequenced](https://github.com/fsharp/FsAutoComplete/blob/f6d5ddcad9e82eb3cb52e01cd3c86d40a205dabe/test/FsAutoComplete.Tests.Lsp/Utils/ServerTests.fs#L33-L38). (Though moving tests into a nested `testList` doesn't seem to run them in parallel/affect performance? -> linked code doesn't seem to be limiting factor :/ )  

Though there might also some issues with getting the correct notifications when tests are run in parallel (I think I remember an issue with that -- though might also be something outdated. I think notifications are now filter by file. So notifications should be assigned to correct tests.)

<br/>

> Might simply be we're not making an LSP server per test or something else.

We're already sharing a server per "logical group" (for example: per CodeFix -- but not for each individual test inside each CodeFix).  
Though pushing that further outward might be good for speed. I remember just reusing a document (`docTestList`) has some noticeable (but not huge) impact. And a single doc should be way more light than a full LSP server.

We could use just one LSP server for all tests. But we're creating a lot of documents -- which I think get all cached in server? Also notifications from LSP Server to test environment are all cached. So might not actually be faster, but more memory intensive -- and definitely more difficult to debug. 
Also: some tests require extra settings -> must get their own LSP server.  

I think keeping the existing "One LSP server per logical group" makes most sense: State limited to logical context. And LSP settings explicit for that context.  
(Though some behind the scene server caching and clearing a server state instead of throwing away might be reasonable too?)

<br/>

> I do vaguely remember the "wait for parse/check notifications" is how a lot of tests know how to continue after requesting a typecheck which can fail and not provide the correct feedback or gets lost in the logging.

[Yes](https://github.com/fsharp/FsAutoComplete/blob/f6d5ddcad9e82eb3cb52e01cd3c86d40a205dabe/test/FsAutoComplete.Tests.Lsp/Utils/Server.fs#L242-L321). Inclusive a required short delay even when correct notification already arrived ... just to ensure it's really the latest notification. Over 100s and 1000s of tests this adds up...  

I think getting rid of this notification stream would be the best for tests: Difficult to filter & debug, might be out of order, lots of waiting to ensure/hope correct messages, lots of caching of old messages.  
Replacing this with direct calls and direct returns would be great -- but then also not the LSP way :/

Reply:
> I think it's just the initialization and destruction of LSP server that must be sequential. The tests themselves should be able to run in parallel -- but are currently in [sequenced](https://github.com/fsharp/FsAutoComplete/blob/f6d5ddcad9e82eb3cb52e01cd3c86d40a205dabe/test/FsAutoComplete.Tests.Lsp/Utils/ServerTests.fs#L33-L38). (Though moving tests into a nested testList doesn't seem to run them in parallel/affect performance? -> linked code doesn't seem to be limiting factor :/ )

> We're already sharing a server per "logical group" (for example: per CodeFix -- but not for each individual test inside each CodeFix).
Though pushing that further outward might be good for speed. I remember just reusing a document (docTestList) has some noticeable (but not huge) impact. And a single doc should be way more light than a full LSP server.

Yeah it would be something to test. If an LSP server per test is more stable, is it worth the extra 10% in time it takes to do the init/shutdown dance? Hard to say unless we try it out.  Although I don't think it's our biggest source of slowness. I think sequenced combined with my findings below is the painful combination.

> [Yes](https://github.com/fsharp/FsAutoComplete/blob/f6d5ddcad9e82eb3cb52e01cd3c86d40a205dabe/test/FsAutoComplete.Tests.Lsp/Utils/Server.fs#L242-L321). Inclusive a required short delay even when correct notification already arrived ... just to ensure it's really the latest notification. Over 100s and 1000s of tests this adds up...
>
>I think getting rid of this notification stream would be the best for tests: Difficult to filter & debug, might be out of order, lots of waiting to ensure/hope correct messages, lots of caching of old messages.
Replacing this with direct calls and direct returns would be great -- but then also not the LSP way :/

Yeah the biggest slowness I think is the `|> Observable.bufferSpan (timeout)` as we have to wait for that full timeout.  It would probably be better to have to pass in your desired criteria and only wait as long as we need. Something like `Observable.choose ... |> Observable.timeout`. This would probably give another big boost in speeding things up.  

We could move to some [pull based diagnostics](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_pullDiagnostics) and poll until we get what we need but I fear we might end up not "truly" testing how an LSP is used from a client.

Reply:
> Might be worth switching to RoslynSourceText by default then eventually removing NamedText

The rest of the discussion is too long, didn't read... but we should do that, and also switch to the Adaptive server, and then remove the old version.